monitoring/ping.go

```diff
@@ -0,0 +1,306 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/satellite/agent"
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/codahale/hdrhistogram"
+	"github.com/gravitational/trace"
+	"github.com/gravitational/ttlmap"
+	serf "github.com/hashicorp/serf/client"
+	log "github.com/sirupsen/logrus"
+)
+
+// TODO: latencyThreshold should be configurable
+// TODO: latency stats should be sent to metrics
+
+const (
+	// pingCheckerID specifies the check name
+	pingCheckerID = "ping-checker"
+	// latencyStatsTTL specifies how long check results will be kept before being dropped
+	latencyStatsTTL = 1 * time.Hour
+	// latencyStatsCapacity sets the number of TTLMaps that can be stored; this will be the size of the cluster -1
+	latencyStatsCapacity = 1000
+	// latencyStatsSlidingWindowSize specifies the number of retained check results
+	latencyStatsSlidingWindowSize = 20
+	// pingMinimum sets the minimum value that can be recorded
+	pingMinimum = 0 * time.Second
+	// pingMaximum sets the maximum value that can be recorded
+	pingMaximum = 10 * time.Second
+	// pingSignificantFigures specifies how many decimals should be recorded
+	pingSignificantFigures = 3
+	// latencyThreshold sets the RTT threshold
+	latencyThreshold = 15 * time.Millisecond
+	// latencyQuantile sets the quantile used while checking Histograms against RTT results
+	latencyQuantile = 95.0
+)
+
+// pingChecker is a checker that verifies that ping times (RTT) between nodes in
+// the cluster are within a predefined threshold
+type pingChecker struct {
+	self           serf.Member
+	serfClient     agent.SerfClient
+	serfRPCAddr    string
+	serfMemberName string
+	latencyStats   ttlmap.TTLMap
+	mux            sync.Mutex
+	logger         log.Entry
+}
+
+// PingCheckerConfig is used to store all the configuration related to the current check
+type PingCheckerConfig struct {
+	// SerfRPCAddr is the address used by the Serf RPC client to communicate
+	SerfRPCAddr string
+	// SerfMemberName is the name associated to this node in Serf
+	SerfMemberName string
+	// NewSerfClient is an optional Serf Client function that can be used instead
+	// of the default one. If not specified it will fallback to the default one
+	NewSerfClient agent.NewSerfClientFunc
+}
+
+// CheckAndSetDefaults is an helper function which just check that the provided
```

Comment:
// CheckAndSetDefaults validates that this configuration is correct and sets value defaults where necessary